components/contract-functions/contract-function.tsx

```diff
@@ -144,27 +152,78 @@ interface ContractFunctionsPanelProps {
   contract?: SmartContract;
 }
 
+type ExtensionFunctions = {
+  extension: string;
+  functions: AbiFunction[];
+};
+
 export const ContractFunctionsPanel: React.FC<ContractFunctionsPanelProps> = ({
   fnsOrEvents,
   contract,
 }) => {
+  const extensions = useContractEnabledExtensions(contract?.abi);
   const isFunction = "stateMutability" in fnsOrEvents[0];
+  const functionsWithExtension = useMemo(() => {
+    let allFunctions = fnsOrEvents as AbiFunction[];
+    const results: ExtensionFunctions[] = [];
+    const processedFunctions: string[] = [];
+    extensions.forEach((ext) => {
+      let functions = extractFunctionsFromAbi(joinABIs(ext.abis as any));
+      allFunctions = allFunctions.filter(
+        (fn) => !functions.map((f) => f.name).includes(fn.name),
+      );
+      functions = functions.filter(
+        (fn) => !processedFunctions.includes(fn.name),
+      );
+      processedFunctions.push(...functions.map((fn) => fn.name));
+      results.push({
+        extension: ext.name as string,
+        functions,
+      });
+    });
+    results.push({
+      extension: "",
+      functions: allFunctions,
+    });
+    return results;
+  }, [fnsOrEvents, extensions]);
 
-  const writeFunctions: AbiFunction[] = useMemo(() => {
-    return fnsOrEvents.filter(
-      (fn) =>
-        (fn as AbiFunction).stateMutability !== "pure" &&
-        (fn as AbiFunction).stateMutability !== "view" &&
-        "stateMutability" in fn,
-    ) as AbiFunction[];
-  }, [fnsOrEvents]);
-  const viewFunctions: AbiFunction[] = useMemo(() => {
-    return fnsOrEvents.filter(
-      (fn) =>
-        (fn as AbiFunction).stateMutability === "pure" ||
-        (fn as AbiFunction).stateMutability === "view",
-    ) as AbiFunction[];
-  }, [fnsOrEvents]);
+  const writeFunctions: ExtensionFunctions[] = useMemo(() => {
+    return functionsWithExtension
+      .map((e) => {
+        const filteredFunctions = e.functions.filter(
+          (fn) =>
+            (fn as AbiFunction).stateMutability !== "pure" &&
+            (fn as AbiFunction).stateMutability !== "view",
+        );
+        if (filteredFunctions.length === 0) {
+          return undefined;
+        }
+        return {
+          extension: e.extension,
+          functions: filteredFunctions,
+        };
+      })
+      .filter((e) => e !== undefined) as ExtensionFunctions[];
+  }, [functionsWithExtension]);
```

Comment:
nit: line breaks between functions. will add to lint rules

Reply:
prettier did this though